# Data Cleaning

In this file, I import the data from the Yelp dataset on Kaggle and clean it to use for the other parts of my project. The original dataset contains the following JSON files. Here are the ones I used for this project:

**1.** **business.json**

Contains business data including location data, attributes, and categories.

**2. review.json**

Contains full review text data including the user_id that wrote the review and the business_id the review is written for.

**3. user.json**

User data including the user's friend mapping and all the metadata associated with the user.

In [ ]:
# Importing Packages

import json
import csv
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from google.colab import files
from tqdm import tqdm, tqdm_notebook, tnrange

## Looking at the Business Dataset


In [ ]:
#Understanding the categories within "Business"
business_arr = []

with open("/content/drive/MyDrive/Yelp_Dataset/business.json") as f:
    for line in f:
        business_arr.append(json.loads(line))

# Viewing the vector
business_arr[0]

{'business_id': '1SWheh84yJXfytovILXOAQ',
 'name': 'Arizona Biltmore Golf Club',
 'address': '2818 E Camino Acequia Drive',
 'city': 'Phoenix',
 'state': 'AZ',
 'postal_code': '85016',
 'latitude': 33.5221425,
 'longitude': -112.0184807,
 'stars': 3.0,
 'review_count': 5,
 'is_open': 0,
 'attributes': {'GoodForKids': 'False'},
 'categories': 'Golf, Active Life',
 'hours': None}

In [ ]:
#Turning array into pd
business = {"business_id": [], "name": [], "address": [], "city": [], "state": [], "postal_code": [], "latitude": [], "longitude": [], "stars": [], "review_count": [], "is_open": [], "attributes": [], "categories": [], "hours": []}
for index in range(0, len(business_arr)):
  business['business_id'].append(business_arr[index]['business_id'])
  business['name'].append(business_arr[index]['name'])
  business['address'].append(business_arr[index]['address'])
  business['city'].append(business_arr[index]['city'])
  business['state'].append(business_arr[index]['state'])
  business['postal_code'].append(business_arr[index]['postal_code'])
  business['latitude'].append(business_arr[index]['latitude'])
  business['longitude'].append(business_arr[index]['longitude'])
  business['stars'].append(business_arr[index]['stars'])
  business['review_count'].append(business_arr[index]['review_count'])
  business['is_open'].append(business_arr[index]['is_open'])
  business['attributes'].append(business_arr[index]['attributes'])
  business['categories'].append(business_arr[index]['categories'])
  business['hours'].append(business_arr[index]['hours'])


business_df = pd.DataFrame(business)
business_df.head()


business_id                             name  \
0  1SWheh84yJXfytovILXOAQ       Arizona Biltmore Golf Club   
1  QXAEGFB4oINsVuTFxEYKFQ       Emerald Chinese Restaurant   
2  gnKjwL_1w79qoiV3IC_xQQ      Musashi Japanese Restaurant   
3  xvX2CttrVhyG2z1dFg_0xw  Farmers Insurance - Paul Lorenz   
4  HhyxOkGAM07SRYtlQ4wMFQ              Queen City Plumbing   

                          address         city state postal_code   latitude  \
0     2818 E Camino Acequia Drive      Phoenix    AZ       85016  33.522143   
1            30 Eglinton Avenue W  Mississauga    ON     L5R 3E7  43.605499   
2       10110 Johnston Rd, Ste 15    Charlotte    NC       28210  35.092564   
3   15655 W Roosevelt St, Ste 237     Goodyear    AZ       85338  33.455613   
4  4209 Stuart Andrew Blvd, Ste F    Charlotte    NC       28217  35.190012   

    longitude  stars  review_count  is_open  \
0 -112.018481    3.0             5        0   
1  -79.652289    2.5           128        1   
2  -80.859132    4.0           170        1   
3 -112.395596    5.0             3        1   
4  -80.887223    4.0             4        1   

                                          attributes  \
0                           {'GoodForKids': 'False'}   
1  {'RestaurantsReservations': 'True', 'GoodForMe...   
2  {'GoodForKids': 'True', 'NoiseLevel': 'u'avera...   
3                                               None   
4  {'BusinessAcceptsBitcoin': 'False', 'ByAppoint...   

                                          categories  \
0                                  Golf, Active Life   
1  Specialty Food, Restaurants, Dim Sum, Imported...   
2                  Sushi Bars, Restaurants, Japanese   
3                      Insurance, Financial Services   
4  Plumbing, Shopping, Local Services, Home Servi...   

                                               hours  
0                                               None  
1  {'Monday': '9:0-0:0', 'Tuesday': '9:0-0:0', 'W...  
2  {'Monday': '17:30-21:30', 'Wednesday': '17:30-...  
3  {'Monday': '8:0-17:0', 'Tuesday': '8:0-17:0', ...  
4  {'Monday': '7:0-23:0', 'Tuesday': '7:0-23:0', ...

In [ ]:
#Filtering Dataset

food_categories = business_df['categories'].str.split(', ').explode().value_counts().head(30)
food_categories

categories
Restaurants                  59371
Shopping                     31878
Food                         29989
Home Services                19729
Beauty & Spas                19370
Health & Medical             17171
Local Services               13932
Automotive                   13203
Nightlife                    13095
Bars                         11341
Event Planning & Services    10371
Active Life                   9521
Fashion                       7798
Sandwiches                    7332
Coffee & Tea                  7321
Fast Food                     7257
American (Traditional)        7107
Hair Salons                   6955
Pizza                         6804
Home & Garden                 6489
Arts & Entertainment          6304
Professional Services         6276
Auto Repair                   6140
Hotels & Travel               6033
Doctors                       5867
Real Estate                   5677
Burgers                       5404
Breakfast & Brunch            5381
Nail Salons                   5043
Specialty Food                4883
Name: count, dtype: int64

I am interested in looking at **Restaurants, Food, Breakfast & Brunch, Fast Food**, so in the next few steps, I am going to be filtering my dataset to include businesses with those categories.



In [ ]:
food_business = business_df[business_df["categories"].str.contains("Restaurants")|business_df["categories"].str.contains("Food")|business_df["categories"].str.contains("Breakfast & Brunch")|business_df["categories"].str.contains("Fast Food")]
food_business.head()

business_id                         name  \
1   QXAEGFB4oINsVuTFxEYKFQ   Emerald Chinese Restaurant   
2   gnKjwL_1w79qoiV3IC_xQQ  Musashi Japanese Restaurant   
11  1Dfx3zM-rW4n-31KeC8sJg                    Taco Bell   
13  fweCYi8FmbJXHCqLnwuk8w                Marco's Pizza   
14  -K4gAv8_vjx8-2BxkVeRkA                   Baby Cakes   

                      address                city state postal_code  \
1        30 Eglinton Avenue W         Mississauga    ON     L5R 3E7   
2   10110 Johnston Rd, Ste 15           Charlotte    NC       28210   
11    2450 E Indian School Rd             Phoenix    AZ       85016   
13            5981 Andrews Rd  Mentor-on-the-Lake    OH       44060   
14               4145 Erie St          Willoughby    OH       44094   

     latitude   longitude  stars  review_count  is_open  \
1   43.605499  -79.652289    2.5           128        1   
2   35.092564  -80.859132    4.0           170        1   
11  33.495194 -112.028588    3.0            18        1   
13  41.708520  -81.359556    4.0            16        1   
14  41.639860  -81.406396    3.0             7        1   

                                           attributes  \
1   {'RestaurantsReservations': 'True', 'GoodForMe...   
2   {'GoodForKids': 'True', 'NoiseLevel': 'u'avera...   
11  {'RestaurantsTakeOut': 'True', 'BusinessParkin...   
13  {'RestaurantsPriceRange2': '2', 'BusinessAccep...   
14  {'RestaurantsTakeOut': 'True', 'BusinessParkin...   

                                           categories  \
1   Specialty Food, Restaurants, Dim Sum, Imported...   
2                   Sushi Bars, Restaurants, Japanese   
11  Restaurants, Breakfast & Brunch, Mexican, Taco...   
13         Italian, Restaurants, Pizza, Chicken Wings   
14                                     Bakeries, Food   

                                                hours  
1   {'Monday': '9:0-0:0', 'Tuesday': '9:0-0:0', 'W...  
2   {'Monday': '17:30-21:30', 'Wednesday': '17:30-...  
11  {'Monday': '7:0-0:0', 'Tuesday': '7:0-0:0', 'W...  
13  {'Monday': '10:0-0:0', 'Tuesday': '10:0-0:0', ...  
14  {'Tuesday': '11:0-17:0', 'Wednesday': '11:0-17...

In [ ]:
#Look at top states and cities
print("Top States")
top_states = food_business['state'].value_counts().head(10)
top_states

Top States


state
ON    18386
AZ    14785
NV    10061
OH     6777
QC     6651
NC     5550
PA     5061
AB     3845
WI     2133
IL      813
Name: count, dtype: int64

In [ ]:
print("Top Cities")
top_cities = food_business['city'].value_counts().head(25)
top_cities

Top Cities


city
Toronto            10093
Las Vegas           8286
Phoenix             5130
Montréal            4540
Calgary             3694
Charlotte           3491
Pittsburgh          3124
Scottsdale          2004
Cleveland           1813
Mississauga         1732
Mesa                1504
Madison             1393
Tempe               1342
Henderson           1171
Chandler            1120
Markham             1055
Glendale             912
Gilbert              784
Brampton             624
Scarborough          613
Richmond Hill        606
North York           573
Vaughan              569
Champaign            509
North Las Vegas      482
Name: count, dtype: int64

In this dataset, we can see that the city with the most number of restaurants is Toronto. I am from Canada (specifically ON), so I am interested in seeing restaurants in **Ontario**.

In [ ]:
ontario_df = food_business[food_business['state'].str.contains('ON')]
ontario_df.shape

(18387, 14)

After filtering out the categories I want, the final dataset is 18,387 values. We have **18,387** rows of data for **Restaurant and Food Businesses in Ontario**.

In [ ]:
#Picking Columns
filtered_restaurants_df = ontario_df.drop(columns = ["hours"])
filtered_restaurants_df.isnull().sum()

business_id       0
name              0
address           0
city              0
state             0
postal_code       0
latitude          0
longitude         0
stars             0
review_count      0
is_open           0
attributes      920
categories        0
dtype: int64

In [ ]:
#Getting rid of null values
filtered_restaurants_df.dropna(subset=['attributes'], inplace=True)
ontario_df.isnull().sum()

business_id        0
name               0
address            0
city               0
state              0
postal_code        0
latitude           0
longitude          0
stars              0
review_count       0
is_open            0
attributes       920
categories         0
hours           4761
dtype: int64

## Looking at Review Dataset

In [ ]:
# Understanding the categories within "Review"
valid_business_ids = set(filtered_restaurants_df["business_id"])

# Initialize a list to store filtered reviews
filtered_reviews = []

# Process the reviews dataset line by line and filter
with open("/content/drive/MyDrive/Yelp_Dataset/review.json") as f:
    for line in f:
        review = json.loads(line)
        # Check if both user_id and business_id are valid
        if review["business_id"] in valid_business_ids:
            filtered_reviews.append({
                "review_id": review["review_id"],
                "user_id": review["user_id"],
                "business_id": review["business_id"],
                "stars": review["stars"],
                "date": review["date"],
                "text": review["text"],
                "useful": review["useful"],
                "funny": review["funny"],
                "cool": review["cool"]
            })

filtered_reviews_df = pd.DataFrame(filtered_reviews)

print(filtered_reviews_df.shape)
filtered_reviews_df.head()

(632227, 9)


review_id                 user_id             business_id  \
0  G7XHMxG0bx9oBJNECG4IFg  jlu4CztcSxrKx56ba1a5AQ  3fw2X5bZYeW9xCz_zGhOHg   
1  JVcjMhlavKKn3UIt9p9OXA  TpyOT5E16YASd7EWjLQlrw  AakkkTuGZA2KBodKi2_u8A   
2  svK3nBU7Rk8VfGorlrN52A  NJlxGtouq06hhC7sS2ECYw  YvrylyuWgbP90RgMqZQVnQ   
3  4bUyL7lzoWzDZaJETAKREg  _N7Ndn29bpll_961oPeEfw  y-Iw6dZflNix4BdwIyTNGA   
4  E6B-2U2sGG3xgmnNWZAEew  DbccYu3OppWKl21OanZnTg  YSUcHqlKMPHHJ_cTrqtNrA   

   stars                 date  \
0    3.0  2016-05-07 01:21:02   
1    1.0  2012-07-16 00:37:14   
2    5.0  2017-04-07 21:27:49   
3    3.0  2014-06-27 21:19:23   
4    1.0  2017-12-29 13:55:19   

                                                text  useful  funny  cool  
0  Tracy dessert had a big name in Hong Kong and ...       5      4     5  
1  I cannot believe how things have changed in 3 ...       1      1     0  
2  You can't really find anything wrong with this...       0      0     0  
3  Good selection of classes of beers and mains. ...       0      0     0  
4  Came here on a Thursday night at 6:30 p.m. My ...       0      0     0

In [ ]:
# Checking for Null Values

filtered_reviews_df.isnull().sum()

review_id      0
user_id        0
business_id    0
stars          0
date           0
text           0
useful         0
funny          0
cool           0
dtype: int64

## Looking at User Dataset

In [ ]:
# Initialize an empty dictionary to hold the filtered data
# Initialize a dictionary to store filtered user data
users = {
    "user_id": [],
    "name": [],
    "review_count": [],
    "yelping_since": [],
    "useful": [],
    "funny": [],
    "cool": [],
    "friends": [],
    "fans": [],
    "average_stars": [],
}

# Extract valid user_ids from the filtered reviews DataFrame
valid_user_ids = set(filtered_reviews_df["user_id"])

# Read and process the user dataset line by line
with open("/content/drive/MyDrive/Yelp_Dataset/user.json") as f:
    for line in f:
        user = json.loads(line)  # Parse JSON line
        # Filter users with review_count >= 10 and valid user_id
        if user["review_count"] >= 10 and user["user_id"] in valid_user_ids:
            users["user_id"].append(user["user_id"])
            users["name"].append(user["name"])
            users["review_count"].append(user["review_count"])
            users["yelping_since"].append(user["yelping_since"])
            users["useful"].append(user["useful"])
            users["funny"].append(user["funny"])
            users["cool"].append(user["cool"])
            users["friends"].append(user["friends"])
            users["fans"].append(user["fans"])
            users["average_stars"].append(user["average_stars"])

# Convert the filtered user data into a pandas DataFrame
filtered_users_df = pd.DataFrame(users)

# Display the shape and head of the DataFrame
print(filtered_users_df.shape)
filtered_users_df.head()

(40086, 10)


user_id         name  review_count        yelping_since  \
0  gvXtMj3XuPr0xHjgmlmtng        Peter            47  2014-01-05 20:45:54   
1  pU6GoRTcl1rIOi6zMynjog       Javier           470  2011-12-16 16:13:01   
2  _ICqwy7SW8J6OJBab1HkwA      allison           391  2007-10-06 02:42:28   
3  rnU1EFMSnVJGrQxrnjaODg         Fred           431  2010-12-28 01:37:20   
4  AsYMm_O4H-mwmWbmeACDHw  Christopher           318  2015-01-17 17:04:09   

   useful  funny  cool                                            friends  \
0      57     26    34  CfGCj80EdA-xS-mTWlAn4Q, JgD2Rk9K07MkZgG7Nb9YzA...   
1    1655   1007   974  tRC9YLo4LHquMVXZ9VO4Ag, -YpfTgz88rsPwsOvlfKn7w...   
2     708    316   399  o8H8CioXzL4ZR_zxZYcv3Q, yJB1hHDIq18Qvmo3yo9_KA...   
3    3157   1338  2158  HHmX2L4Gg6L0WJNcxs4Ptw, 8DEyKVyplnOcSKx39vatbg...   
4     400    175   223  4I_AqfW1E_OLDskrgxxU0Q, 0hm1__RgRWYvHXxWfUJMyA...   

   fans  average_stars  
0     9           3.60  
1    71           3.45  
2    13           4.31  
3    58           3.95  
4    24           3.64

In [ ]:
#Checking for Null Values
filtered_users_df.isnull().sum()

user_id          0
name             0
review_count     0
yelping_since    0
useful           0
funny            0
cool             0
friends          0
fans             0
average_stars    0
dtype: int64

## Final Datasets

In [ ]:
from google.colab import files

files.download("filtered_restaurants.csv")
files.download("filtered_reviews.csv")
files.download("filtered_users.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>